src/Tools/Source/CompilerGeneratorTools/Source/VisualBasicSyntaxGenerator/Grammar/GrammarGenerator.vb

```diff
@@ -0,0 +1,295 @@
+        _majorRules = ImmutableArray.Create(
+            "CompilationUnitSyntax", "StatementSyntax", "ExpressionSyntax", "TypeSyntax", "NameSyntax", "XmlNodeSyntax", "StructuredTriviaSyntax", "Modifier", "SyntaxToken", "PunctuationSyntax")
+
+        Dim result = "// <auto-generated />" + Environment.NewLine + "grammar vb;" + Environment.NewLine
+
+        ' Handle each major section first And then walk any rules Not hit transitively from them.
+        For Each rule In _majorRules.Concat(_nameToProductions.Keys.OrderBy(Function(a) a))
+            ProcessRule(rule, result)
+        Next
+
+        Return result
+    End Function
+
+    Private Shared Function Join(delim As String, productions As IEnumerable(Of Production)) As Production
+        Return New Production(
+            String.Join(delim, productions.Where(Function(p) p.Text.Length > 0)),
+            productions.SelectMany(Function(p) p.ReferencedRules))
+    End Function
+
+    Private Function HandleChildren(structureNode As ParseNodeStructure, children As List(Of ParseNodeChild), Optional delim As String = " ") As Production
+        Return Join(delim, children.Select(
+                    Function(child) HandleField(structureNode, child)))
+    End Function
+
+    Private Function HandleField(structureNode As ParseNodeStructure, child As ParseNodeChild) As Production
+        If child.IsSeparated Then
+            Return HandleSeparatedList(structureNode, child)
+        ElseIf child.IsList Then
+            Return HandleList(structureNode, child)
+        Else
+            Return HandleChildKind(structureNode, child, child.ChildKind).Suffix("?", [when]:=child.IsOptional)
+        End If
+    End Function
+
+    Private Function HandleSeparatedList(structureNode As ParseNodeStructure, child As ParseNodeChild) As Production
+        Dim childProduction = HandleChildKind(structureNode, child, child.ChildKind)
+        Dim separatorProd = HandleChildKind(structureNode, child, child.SeparatorsKind)
+
+        Return childProduction.Suffix("(" + separatorProd.Text + " " + childProduction.Text + ")*").Parenthesize().Suffix("?")
+    End Function
+
+    Private Function HandleList(structureNode As ParseNodeStructure, child As ParseNodeChild) As Production
+        Return HandleChildKind(structureNode, child, child.ChildKind).Suffix("*")
+    End Function
+
+    Private Function HandleChildKind(structureNode As ParseNodeStructure,
+                                     child As ParseNodeChild,
+                                     childKind As Object) As Production
+        If child.Name = "Modifiers" Then
+            Return RuleReference("Modifier")
+        End If
+
+        Dim x = child.ChildKind()
+        Dim y = child.ChildKind("")
+
+        Dim nodeKind = TryCast(childKind, ParseNodeKind)
+        Dim nodeKindList = TryCast(childKind, List(Of ParseNodeKind))
+
+        If structureNode.NodeKinds IsNot Nothing Then
```

Comment:
this is one of the more complex areas in the VB code versus the C# code, and it's likely this needs tweaking.  the main issue is how many VB types have an abstract base type (like 'Block') which define the abstract shape of `begin_statement statements* end_statement` which then makes a relationship between the keyword of the block and the kind of the block (and it's subclasses).  This info is def encoded in the VB syntax model, but i'm clearly not quite extracting it properly.